### ISSUE TITLE
is there support for asp.net core 2.0

### ISSUE THREAD

**Issue by darewreck54** [Microsoft/PowerBI-CSharp#116]
Hi,

Wondering if there is support for the asp.net core 2.0 for this lib?

Thanks,
Derek

**Comment by ali-hamud** [comment]
Hi @darewreck54,
Thanks for opening this issue.
.Net Core 2.0 has been announced 3 days ago, so we didn't have a chance to check this yet.

No ETA.

**Comment by wind-rider** [comment]
@ali-hamud I really appreciate your work on this PowerBI client; however from a Microsoft employee I would expect a different answer. The previews for .net core and asp.net core 2.0 are out for months already so of course you had a chance to check it out; other than that there never came an answer to questions about PowerBI support for use with .net core and asp.net core 1.0... .Net Core 1.0 is out for more than a year now and questions about it seemed to be ignored.

Please see https://github.com/Microsoft/PowerBI-CSharp/issues/30, https://github.com/Microsoft/PowerBI-Developer-Samples/issues/3 and https://github.com/Microsoft/PowerBI-CSharp/issues/95. It feels as if users are talking to a wall. I don't like giving feedback this way in public, but I don't see where I can do it privately.

**Comment by ali-hamud** [comment]
.Net Core 1.0 is supported starting from version 2.0.0 of Microsoft.PowerBI.API.

For Server side (management part) please use Microsoft.PowerBI.API package.
https://www.nuget.org/packages/Microsoft.PowerBI.Api/

For client side (embed part) please use bower package of javascript library:
https://github.com/Microsoft/PowerBI-javascript

Actually, I think it should work the same for .Net Core 2.0
But, as I didn't try it I cannot promise.

**Comment by ali-hamud** [comment]
And in addition, last time I checked, non-interactive sign-in is not supported in dotnet core.
So, embedding will work for interactive sign-in with Power BI User only.

**Comment by wind-rider** [comment]
A possible way could be using client certificate authentication - I got this answer in the ADAL GitHub repo (see https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/issues/482#issuecomment-316973560)
I haven’t yet been able to try it though. [Somebody tried](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/issues/482#issuecomment-323574193) but didn't get it to work, maybe somebody else will?

**Comment by ali-hamud** [comment]
Hello @wind-rider 
I passed the issue to the relevant team. I will update when I have an answer.
No ETA.

**Comment by darewreck54** [comment]
Both projects do not seem to be fully compatible with .net core 2.  
![image](https://user-images.githubusercontent.com/16171726/29534004-2e152d52-8669-11e7-92eb-7dd6e27d0dec.png)

**Comment by nicksav** [comment]
Update?

**Comment by nicksav** [comment]
Haven't tested all methods, but GetReports works. So, what needs to be done. First, .net core dosnt support username/pass for requesting a token. However, you can simply do it by making a HTTP request to auth2 endpoint:

`
                var tokenEndpoint = @"https://login.windows.net/{tenantid}/oauth2/token";
                var accept = "application/json";

                client.DefaultRequestHeaders.Add("Accept", accept);
                string postBody = @"resource=https%3A%2F%2Fanalysis.windows.net/powerbi/api
                                    &client_id={clientId}
                                    &grant_type=password
                                    &username=username
                                    &password=password
                                    &scope=openid";


                var result = await client.PostAsync(tokenEndpoint, new StringContent(postBody, Encoding.UTF8, "application/x-www-form-urlencoded")).ContinueWith<string>((response) =>
                {
                    return response.Result.Content.ReadAsStringAsync().Result;
                });

                JObject jobject = JObject.Parse(result);

                var token = jobject["access_token"].Value<string>();

                var tokenCredentials = new TokenCredentials(token, "Bearer");
`

This should give you a token. 

The second thing, you need to create a brand new class library project targeting .net core. Copy 2 folders for Imports and Source. Add latest package for Microsoft.Rest.ClientRuntime and build. Reference the lib in your .net core project and use it:

`
                var tokenCredentials = new TokenCredentials(token, "Bearer");

                // Create a Power BI Client object. It will be used to call Power BI APIs.
                using (var cl = new PowerBIClient(new Uri(ApiUrl), tokenCredentials))
                {
                    // Get the first report all reports in that workspace
                    ODataResponseListReport reports = cl.Reports.GetReports();
                    Report report = reports.Value.FirstOrDefault();
                    var embedUrl = report.EmbedUrl;
                }
`

**Comment by AlwaysAbhl001** [comment]
@ali-hamud  any update?

**Comment by Jan-Olof** [comment]
@nicksav I tried your HTTP request to the auth2 endpoint. However I only get "invalid_grant" back. ("error_description": "AADSTS70002: Error validating credentials. AADSTS50126: Invalid username or password)

**Comment by nicksav** [comment]
@Jan-Olof Actually got all methods working and just about to put to production.
You still need to create an App in Azure DB and also Grant Permissions.

Thanks
Nick

**Comment by Jan-Olof** [comment]
@nicksav Yes, I got everything to work with the old sample. Only the HTTP request you described returns the "invalid_grant". I wonder why? Good to know that you got it working though.

**Comment by BennyM** [comment]
I use the plain API in .net core. Works just as good, the SDK is a real thin layer around it.

**Comment by BennyM** [comment]
I have ported the library to .NET Standard 2.0 and published to NuGet. 

- Source & Instructions: https://github.com/BennyM/PowerBI-CSharp
- Package: https://www.nuget.org/packages/PowerBI.Api/

**Comment by ali-hamud** [comment]
Hi @BennyM 
I don't see how you encrypt the password before sending over the network, which makes the solution insecure. please correct me if I am wrong.

@Jan-Olof 
you may missed the Grant Permission part. sign in to Azure, go to app registration to the app you created, make sure you requested permissions to the needed Power BI APIs and click Grant Permissions.

@AlwaysAbhl001 
Unfortunately I have no updates yet.

I recommend to have a .NET Full service which is responsible for generating tokens, and user Service to Service communication in your .Net Core project to get embed token.

**Comment by BennyM** [comment]
@ali-hamud It's as secure as the UserPasswordCredential  in the current full .NET api. You should not dismiss it that easily. It simulates a user and it's send over HTTPS thus encrypted. It's also the only way to create embed tokens as mentioned here: https://blogs.msdn.microsoft.com/tsmatsuz/2017/07/14/power-bi-embed-report-in-app/. As long as Power BI does not support app only tokens both the official SDK and my unofficial port suffer this.

**Comment by ali-hamud** [comment]
@BennyM 
True. I missed the HTTPS encryption.
I took a look on UserPasswordCredential implementation and It's also has no encryption so you are ok. implementation is here:
https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/3982cb2886d8b6e561d890e749eb932a55439145/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/net45/UserPasswordCredential.cs

They have SecureString class which deletes the string from memory once you done with it, but no encryption. so, you are ok.

**Comment by Jan-Olof** [comment]
I still can't get the GetAccessToken call directly to the api (that replaces the UserPasswordCredential) to work with our Power BI Pro users. Using the current full .NET api it works fine. Really strange since it works for both @nicksav and @BennyM. I tried using the code provided by @nicksav and @BennyM as well as using Postman but it simply returns "invalid_grant" as stated above.

I ended up using the full .NET api in an Azure Function for this. At least we now have a working embedded solution.
